File: remote/src/main/java/com/kelsos/mbrc/mappers/TrackMapper.java
package com.kelsos.mbrc.mappers;

import com.kelsos.mbrc.dao.AlbumDao;
import com.kelsos.mbrc.dao.ArtistDao;
import com.kelsos.mbrc.dao.GenreDao;
import com.kelsos.mbrc.dao.TrackDao;
import com.kelsos.mbrc.dao.views.TrackModelView;
import com.kelsos.mbrc.domain.Track;
import com.kelsos.mbrc.dto.library.TrackDto;
import java.util.List;
import rx.Observable;

public class TrackMapper {
  public static Track map(TrackModelView view) {
    return new Track(view.getId(), view.getArtist(), view.getTitle(), view.getCover());
  }

  public static List<Track> map(List<TrackModelView> dao) {
    return Observable.from(dao).map(TrackMapper::map).toList().toBlocking().first();
  }

  public static List<TrackDao> mapDtos(List<TrackDto> trackDto,
      List<ArtistDao> artists,
      List<GenreDao> genres,
      List<AlbumDao> albums) {
    return Observable.from(trackDto).map(dto -> mapDto(dto, artists, genres, albums)).toList().toBlocking().first();
  }

  public static TrackDao mapDto(TrackDto trackDto,
      List<ArtistDao> artists,
      List<GenreDao> genres,
      List<AlbumDao> albums) {
    TrackDao dao = new TrackDao();
    dao.setId(trackDto.getId());
    dao.setPath(trackDto.getPath());
    dao.setPosition(trackDto.getPosition());
    dao.setTitle(trackDto.getTitle());
    dao.setDisc(trackDto.getDisc());
    dao.setYear(trackDto.getYear());
    dao.setDateAdded(trackDto.getDateAdded());
    dao.setDateDeleted(trackDto.getDateDeleted());
    dao.setDateUpdated(trackDto.getDateUpdated());
    dao.setGenre(MapperUtils.getGenreById(trackDto.getGenreId(), genres));
    dao.setAlbumArtist(MapperUtils.getArtistById(trackDto.getAlbumArtistId(), artists));
    dao.setArtist(MapperUtils.getArtistById(trackDto.getArtistId(), artists));
    dao.setAlbum(MapperUtils.getAlbumById(trackDto.getAlbumId(), albums));
    return dao;
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/mappers/TrackMapper.kt
package com.kelsos.mbrc.mappers

import com.kelsos.mbrc.data.dao.AlbumDao
import com.kelsos.mbrc.data.dao.ArtistDao
import com.kelsos.mbrc.data.dao.GenreDao
import com.kelsos.mbrc.data.dao.TrackDao
import com.kelsos.mbrc.data.views.TrackModelView
import com.kelsos.mbrc.domain.Track
import com.kelsos.mbrc.dto.library.TrackDto

object TrackMapper {
    fun map(view: TrackModelView): Track {
        return Track(view.id, view.artist, view.title, view.cover)
    }

    fun map(dao: List<TrackModelView>): List<Track> {
        return dao.map { map(it) }.toList()
    }

    fun mapDtos(trackDto: List<TrackDto>,
                artists: List<ArtistDao>,
                genres: List<GenreDao>,
                albums: List<AlbumDao>): List<TrackDao> {
        return trackDto.map { dto -> mapDto(dto, artists, genres, albums) }.toList()
    }

    fun mapDto(trackDto: TrackDto,
               artists: List<ArtistDao>,
               genres: List<GenreDao>,
               albums: List<AlbumDao>): TrackDao {
        val dao = TrackDao()
        dao.id = trackDto.id
        dao.path = trackDto.path
        dao.position = trackDto.position
        dao.title = trackDto.title
        dao.disc = trackDto.disc
        dao.year = trackDto.year
        dao.dateAdded = trackDto.dateAdded
        dao.dateDeleted = trackDto.dateDeleted
        dao.dateUpdated = trackDto.dateUpdated
        dao.genre = MapperUtils.getGenreById(trackDto.genreId, genres)
        dao.albumArtist = MapperUtils.getArtistById(trackDto.albumArtistId, artists)
        dao.artist = MapperUtils.getArtistById(trackDto.artistId, artists)
        dao.album = MapperUtils.getAlbumById(trackDto.albumId, albums)
        return dao
    }
}
